lib/kdtree.cpp

```diff
@@ -445,25 +445,23 @@ std::vector<detail::FlatNode> flatten(std::unique_ptr<TreeNode> root) {
             stack.emplace(node.right(), node_index);
             stack.emplace(node.left(), INVALID_INDEX);
         } else {
-            bool is_odd = true;
-            TriangleId first_id;
             const TriangleId* triangle_ids = node.triangle_ids();
-            // fill in pairwise triangles as leaf nodes
-            for (size_t i = 0; i < node.num_tris(); ++i) {
-                if (is_odd) {
-                    first_id = triangle_ids[i];
-                } else {
-                    nodes.emplace_back(first_id, triangle_ids[i]);
-                }
-                is_odd = !is_odd;
+            // create a leaf node for each pair of triangles
+            size_t i = 1;
+            for (; i < node.num_tris(); i += 2) {
+                nodes.emplace_back(triangle_ids[i - 1], triangle_ids[i]);
             }
-            // triangle left? => add leaf node with a single triangle
-            if (!is_odd) {
-                nodes.emplace_back(first_id);
+            // a triangle left? => add leaf node with a single triangle
+            if (i - 1 < node.num_tris()) {
+                nodes.emplace_back(triangle_ids[i - 1]);
+                // in that case we don't need a sentinel node, since a
+                // half-empty node can be used as a sentinel
+            } else {
+                // add inner node as sentinel
+                nodes.emplace_back(Axis::X, 0, 0);
```

Comment:
We can't do constexpr since, the constructor of the class is not constexpr, i.e. the class is not constructible during compile time. However we can introduce just a normal variable. Not sure if it really increases the readability.

Done.